### ISSUE TITLE
Using custom callback to set trainable=true doesn't work when use model.fit()

### ISSUE THREAD

**Issue by jason9075** [tensorflow/tensorflow#47906]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Big Sur 11.2.2
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): tensorflow==2.3.0
- Python version: 3.6.10
- Bazel version (if compiling from source): None
- GCC/Compiler version (if compiling from source): None
- CUDA/cuDNN version: None
- GPU model and memory: None

Hi, I want to train the model with transfer learning. First, I load the pretrain model, then I freeze some of layer at front of model. After several epoch, I want to unfreeze the weight for training. I use callback to write some  snippet to test it. But I found that when I set the trainable to True, the model still not training...

Code:

[Colab link](https://colab.research.google.com/drive/1suX_4qMZ0yyBGaIqmDGFLa6KN0hMK5Jo?usp=sharing)

```
import tensorflow as tf
import tensorflow_datasets as tfds
import os
from tensorflow.python.keras.callbacks import ModelCheckpoint

class UnFreezeWeight(tf.keras.callbacks.Callback):
    def __init__(self, freeze_before_epoch):
        super().__init__()
        self.freeze_before_epoch = freeze_before_epoch

    def on_epoch_begin(self, epoch, logs=None):
        if self.freeze_before_epoch != epoch:
            return

        # Unfreeze all weight.
        print('set trainable to True.')
        for layer in self.model.layers:
            layer.trainable = True


def _normalize_img(img, label):
    img = tf.cast(img, tf.float32) / 255.
    return img, label


def main():
    train_ds = tfds.load('mnist', split='train', as_supervised=True)
    train_ds = train_ds.batch(32)
    train_ds = train_ds.map(_normalize_img)

    valid_ds = tfds.load('mnist', split='test', as_supervised=True)
    valid_ds = valid_ds.batch(32)
    valid_ds = valid_ds.map(_normalize_img)

    model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Dense(10)
    ])
    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

    # Freeze all weight.
    for layer in model.layers:
        layer.trainable = False
        
    # If I unfreeze at epoch index 0, the model will learning.
    # But if I un freeze after epoch index 1, the model won't training.
    freeze = UnFreezeWeight(1)

    model.compile(optimizer='adam',
                  loss=loss_fn,
                  metrics=['accuracy'])

    model.fit(train_ds,
              validation_data=valid_ds,
              callbacks=[freeze],
              epochs=5)


if __name__ == '__main__':
    main()
```

Output:
When unfreeze weight at epoch 2
```
Epoch 1/5
1875/1875 [==============================] - 9s 5ms/step - loss: 2.3907 - accuracy: 0.1362 - val_loss: 2.3740 - val_accuracy: 0.1488
Epoch 2/5
set trainable to True.
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3922 - accuracy: 0.1349 - val_loss: 2.3740 - val_accuracy: 0.1488
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3910 - accuracy: 0.1361 - val_loss: 2.3740 - val_accuracy: 0.1488
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3914 - accuracy: 0.1378 - val_loss: 2.3740 - val_accuracy: 0.1488
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3926 - accuracy: 0.1346 - val_loss: 2.3740 - val_accuracy: 0.1488
```

When unfreeze weight at epoch 1
```
Epoch 1/5
set trainable to True.
1875/1875 [==============================] - 12s 6ms/step - loss: 0.4833 - accuracy: 0.8568 - val_loss: 0.1364 - val_accuracy: 0.9607
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1502 - accuracy: 0.9567 - val_loss: 0.0991 - val_accuracy: 0.9711
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1100 - accuracy: 0.9668 - val_loss: 0.0870 - val_accuracy: 0.9751
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0910 - accuracy: 0.9734 - val_loss: 0.0776 - val_accuracy: 0.9744
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0768 - accuracy: 0.9759 - val_loss: 0.0790 - val_accuracy: 0.9751
```

**Comment by Saduf2019** [comment]
@jason9075 
I ran the code shared please refer to [this gist ]( https://colab.research.google.com/gist/Saduf2019/3fd5ae49f624f7ce0d0372602b805101/untitled565.ipynb)and let us know the expected output.

**Comment by jason9075** [comment]
Hi Saduf2019

I have some problem at line:
```
freeze = UnFreezeWeight(10)
```
If I unfreeze at 0, everything runs as expected.
But if I unfreeze at other epoch index, the val accuracy won't raise.

[colab gist](https://colab.research.google.com/gist/Saduf2019/3fd5ae49f624f7ce0d0372602b805101/untitled565.ipynb)

```
Epoch 1/20
1875/1875 [==============================] - 9s 5ms/step - loss: 2.3982 - accuracy: 0.0804 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 2/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3993 - accuracy: 0.0809 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 3/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3977 - accuracy: 0.0804 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 4/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3957 - accuracy: 0.0788 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 5/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3995 - accuracy: 0.0792 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 6/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.4004 - accuracy: 0.0801 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 7/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3998 - accuracy: 0.0789 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 8/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.4010 - accuracy: 0.0781 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 9/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3972 - accuracy: 0.0790 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 10/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3991 - accuracy: 0.0777 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 11/20
set trainable to True.
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3996 - accuracy: 0.0795 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 12/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3996 - accuracy: 0.0767 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 13/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3981 - accuracy: 0.0794 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 14/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3993 - accuracy: 0.0795 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 15/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3993 - accuracy: 0.0769 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 16/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.4013 - accuracy: 0.0807 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 17/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3984 - accuracy: 0.0808 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 18/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3989 - accuracy: 0.0786 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 19/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.4007 - accuracy: 0.0785 - val_loss: 2.3787 - val_accuracy: 0.0685
Epoch 20/20
1875/1875 [==============================] - 3s 2ms/step - loss: 2.3982 - accuracy: 0.0800 - val_loss: 2.3787 - val_accuracy: 0.0685
```

**Comment by Saduf2019** [comment]
@jason9075 
 Incorrect gist shared in the above comment, please share the colab for which the output is posted above.

**Comment by jason9075** [comment]
@Saduf2019 

Sorry, wrong url...
[this one](https://colab.research.google.com/gist/jason9075/a97443f466a9028f7b2f3ffa57e9586c/untitled565.ipynb)

**Comment by Saduf2019** [comment]
@jason9075
I ran the gist shared on tf 2.3, tf 2.4 and nightly, please refer to [this gist](https://colab.research.google.com/gist/Saduf2019/4d8ef8dc33eb6dac577aed7edbca1032/untitled573.ipynb) and confirm.

**Comment by jason9075** [comment]
Hi, @Saduf2019 

In your gist, the tf2.4 the val_loss remain at 2.4462 and val_accuracy remain at 0.0984 after epoch 6 just show that the model is not learning.
And tf 2.3 and nightly is also remain at same val_loss and val_accuracy.
I expect my model start learning after epoch 6 but is not.

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210531,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/a18f116207bdeb344488e90c5ae51ea1/untitled166.ipynb)..Thanks !

**Comment by jvishnuvardhan** [comment]
@jason9075 This got resolved in `TF2.7`. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/9fc435f9586f4141814ad5e580fef69c/untitled.ipynb).

I updated your code by adding one line to the callback.
```
# Unfreeze all weight.
self.model.make_train_function(force=True)
```

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by jason9075** [comment]
@jvishnuvardhan It works!
Thanks!
